File: release/scripts/startup/bl_operators/image.py
# ##### BEGIN GPL LICENSE BLOCK #####
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software Foundation,
#  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# ##### END GPL LICENSE BLOCK #####

# <pep8-80 compliant>

import bpy
from bpy.types import Operator
from bpy.props import StringProperty


class EditExternally(Operator):
    '''Edit image in an external application'''
    bl_idname = "image.external_edit"
    bl_label = "Image Edit Externally"
    bl_options = {'REGISTER'}

    filepath = StringProperty(
            subtype='FILE_PATH',
            )

    def _editor_guess(self, context):
        import sys

        image_editor = context.user_preferences.filepaths.image_editor

        # use image editor in the preferences when available.
        if not image_editor:
            if sys.platform[:3] == "win":
                image_editor = ["start"]  # not tested!
            elif sys.platform == "darwin":
                image_editor = ["open"]
            else:
                image_editor = ["gimp"]
        else:
            if sys.platform == "darwin":
                # blender file selector treats .app as a folder
                # and will include a trailing backslash, so we strip it.
                image_editor.rstrip('\\')
                image_editor = ["open", "-a", image_editor]
            else:
                image_editor = [image_editor]

        return image_editor

    def execute(self, context):
        import os
        import subprocess

        filepath = self.filepath

        if not filepath:
            self.report({'ERROR'}, "Image path not set")
            return {'CANCELLED'}

        if not os.path.exists(filepath) or not os.path.isfile(filepath):
            self.report({'ERROR'},
                        "Image path %r not found, image may be packed or "
                        "unsaved" % filepath)
            return {'CANCELLED'}

        cmd = self._editor_guess(context) + [filepath]

        try:
            subprocess.Popen(cmd)
        except:
            import traceback
            traceback.print_exc()
            self.report({'ERROR'},
                        "Image editor not found, "
                        "please specify in User Preferences > File")

            return {'CANCELLED'}

        return {'FINISHED'}

    def invoke(self, context, event):
        import os
        try:
            image = context.space_data.image
        except AttributeError:
            self.report({'ERROR'}, "Context incorrect, image not found")
            return {'CANCELLED'}

        if image.packed_file:
            self.report({'ERROR'}, "Image is packed, unpack before editing")
            return {'CANCELLED'}

        filepath = bpy.path.abspath(image.filepath, library=image.library)

        self.filepath = os.path.normpath(filepath)
        self.execute(context)

        return {'FINISHED'}


class SaveDirty(Operator):
    """Save all modified textures"""
    bl_idname = "image.save_dirty"
    bl_label = "Save Dirty"
    bl_options = {'REGISTER', 'UNDO'}

    def execute(self, context):
        unique_paths = set()
        for image in bpy.data.images:
            if image.is_dirty:
                filepath = bpy.path.abspath(image.filepath)
                if "\\" not in filepath and "/" not in filepath:
                    self.report({'WARNING'}, "Invalid path: " + filepath)
                elif filepath in unique_paths:
                    self.report({'WARNING'},
                                "Path used by more then one image: %r" %
                                filepath)
                else:
                    unique_paths.add(filepath)
                    image.save()
        return {'FINISHED'}


class ProjectEdit(Operator):
    """Edit a snapshot of the view-port in an external image editor"""
    bl_idname = "image.project_edit"
    bl_label = "Project Edit"
    bl_options = {'REGISTER'}

    _proj_hack = [""]

    def execute(self, context):
        import os

        EXT = "png"  # could be made an option but for now ok

        for image in bpy.data.images:
            image.tag = True

        # opengl buffer may fail, we can't help this, but best report it.
        try:
            bpy.ops.paint.image_from_view()
        except RuntimeError as err:
            self.report({'ERROR'}, str(err))
            return {'CANCELLED'}

        image_new = None
        for image in bpy.data.images:
            if not image.tag:
                image_new = image
                break

        if not image_new:
            self.report({'ERROR'}, "Could not make new image")
            return {'CANCELLED'}

        filepath = os.path.basename(bpy.data.filepath)
        filepath = os.path.splitext(filepath)[0]
        # fixes <memory> rubbish, needs checking
        # filepath = bpy.path.clean_name(filepath)

        if bpy.data.is_saved:
            filepath = "//" + filepath
        else:
            tmpdir = context.user_preferences.filepaths.temporary_directory
            filepath = os.path.join(tmpdir, "project_edit")

        obj = context.object

        if obj:
            filepath += "_" + bpy.path.clean_name(obj.name)

        filepath_final = filepath + "." + EXT
        i = 0

        while os.path.exists(bpy.path.abspath(filepath_final)):
            filepath_final = filepath + ("%.3d.%s" % (i, EXT))
            i += 1

        image_new.name = bpy.path.basename(filepath_final)
        ProjectEdit._proj_hack[0] = image_new.name

        image_new.filepath_raw = filepath_final  # TODO, filepath raw is crummy
        image_new.file_format = 'PNG'
        image_new.save()

        filepath_final = bpy.path.abspath(filepath_final)

        try:
            bpy.ops.image.external_edit(filepath=filepath_final)
        except RuntimeError as re:
            self.report({'ERROR'}, str(re))

        return {'FINISHED'}


class ProjectApply(Operator):
    """Project edited image back onto the object"""
    bl_idname = "image.project_apply"
    bl_label = "Project Apply"
    bl_options = {'REGISTER'}

    def execute(self, context):
        image_name = ProjectEdit._proj_hack[0]  # TODO, deal with this nicer

        try:
            image = bpy.data.images[image_name]
        except KeyError:
            import traceback
            traceback.print_exc()
            self.report({'ERROR'}, "Could not find image '%s'" % image_name)
            return {'CANCELLED'}

        image.reload()
        bpy.ops.paint.project_image(image=image_name)

        return {'FINISHED'}

File: intern/libmv/intern/image.cc
/*
 * ***** BEGIN GPL LICENSE BLOCK *****
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 * The Original Code is Copyright (C) 2011 Blender Foundation.
 * All rights reserved.
 *
 * Contributor(s): Blender Foundation,
 *                 Sergey Sharybin
 *
 * ***** END GPL LICENSE BLOCK *****
 */

#include "intern/image.h"
#include "intern/utildefines.h"
#include "libmv/tracking/track_region.h"

#include <cassert>
#include <png.h>

using libmv::FloatImage;
using libmv::SamplePlanarPatch;

void libmv_floatImageDestroy(libmv_FloatImage *image) {
  delete [] image->buffer;
}

/* Image <-> buffers conversion */

void libmv_byteBufferToFloatImage(const unsigned char* buffer,
                                  int width,
                                  int height,
                                  int channels,
                                  FloatImage* image) {
  image->Resize(height, width, channels);
  for (int y = 0, a = 0; y < height; y++) {
    for (int x = 0; x < width; x++) {
      for (int k = 0; k < channels; k++) {
        (*image)(y, x, k) = (float)buffer[a++] / 255.0f;
      }
    }
  }
}

void libmv_floatBufferToFloatImage(const float* buffer,
                                   int width,
                                   int height,
                                   int channels,
                                   FloatImage* image) {
  image->Resize(height, width, channels);
  for (int y = 0, a = 0; y < height; y++) {
    for (int x = 0; x < width; x++) {
      for (int k = 0; k < channels; k++) {
        (*image)(y, x, k) = buffer[a++];
      }
    }
  }
}

void libmv_floatImageToFloatBuffer(const FloatImage &image,
                                   float* buffer) {
  for (int y = 0, a = 0; y < image.Height(); y++) {
    for (int x = 0; x < image.Width(); x++) {
      for (int k = 0; k < image.Depth(); k++) {
        buffer[a++] = image(y, x, k);
      }
    }
  }
}

void libmv_floatImageToByteBuffer(const libmv::FloatImage &image,
                                  unsigned char* buffer) {
  for (int y = 0, a= 0; y < image.Height(); y++) {
    for (int x = 0; x < image.Width(); x++) {
      for (int k = 0; k < image.Depth(); k++) {
        buffer[a++] = image(y, x, k) * 255.0f;
      }
    }
  }
}

static bool savePNGImage(png_bytep* row_pointers,
                         int width,
                         int height,
                         int depth,
                         int color_type,
                         const char* file_name) {
  png_infop info_ptr;
  png_structp png_ptr;
  FILE *fp = fopen(file_name, "wb");

  if (fp == NULL) {
    return false;
  }

  /* Initialize stuff */
  png_ptr = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
  info_ptr = png_create_info_struct(png_ptr);

  if (setjmp(png_jmpbuf(png_ptr))) {
    fclose(fp);
    return false;
  }

  png_init_io(png_ptr, fp);

  /* Write PNG header. */
  if (setjmp(png_jmpbuf(png_ptr))) {
    fclose(fp);
    return false;
  }

  png_set_IHDR(png_ptr,
               info_ptr,
               width,
               height,
               depth,
               color_type,
               PNG_INTERLACE_NONE,
               PNG_COMPRESSION_TYPE_BASE,
               PNG_FILTER_TYPE_BASE);

  png_write_info(png_ptr, info_ptr);

  /* Write bytes/ */
  if (setjmp(png_jmpbuf(png_ptr))) {
    fclose(fp);
    return false;
  }

  png_write_image(png_ptr, row_pointers);

  /* End write/ */
  if (setjmp(png_jmpbuf(png_ptr))) {
    fclose(fp);
    return false;
  }

  png_write_end(png_ptr, NULL);
  fclose(fp);

  return true;
}

bool libmv_saveImage(const FloatImage& image,
                     const char* prefix,
                     int x0,
                     int y0) {
  int x, y;
  png_bytep *row_pointers;

  assert(image.Depth() == 1);

  row_pointers = new png_bytep[image.Height()];

  for (y = 0; y < image.Height(); y++) {
    row_pointers[y] = new png_byte[4 * image.Width()];

    for (x = 0; x < image.Width(); x++) {
      if (x0 == x && image.Height() - y0 - 1 == y) {
        row_pointers[y][x * 4 + 0] = 255;
        row_pointers[y][x * 4 + 1] = 0;
        row_pointers[y][x * 4 + 2] = 0;
        row_pointers[y][x * 4 + 3] = 255;
      } else {
        float pixel = image(image.Height() - y - 1, x, 0);
        row_pointers[y][x * 4 + 0] = pixel * 255;
        row_pointers[y][x * 4 + 1] = pixel * 255;
        row_pointers[y][x * 4 + 2] = pixel * 255;
        row_pointers[y][x * 4 + 3] = 255;
      }
    }
  }

  static int image_counter = 0;
  char file_name[128];
  snprintf(file_name, sizeof(file_name),
           "%s_%02d.png",
           prefix, ++image_counter);
  bool result = savePNGImage(row_pointers,
                             image.Width(),
                             image.Height(),
                             8,
                             PNG_COLOR_TYPE_RGBA,
                             file_name);

  for (y = 0; y < image.Height(); y++) {
    delete [] row_pointers[y];
  }
  delete [] row_pointers;

  return result;
}

void libmv_samplePlanarPatchFloat(const float* image,
                                  int width,
                                  int height,
                                  int channels,
                                  const double* xs,
                                  const double* ys,
                                  int num_samples_x,
                                  int num_samples_y,
                                  const float* mask,
                                  float* patch,
                                  double* warped_position_x,
                                  double* warped_position_y) {
  FloatImage libmv_image, libmv_patch, libmv_mask;
  FloatImage *libmv_mask_for_sample = NULL;

  libmv_floatBufferToFloatImage(image, width, height, channels, &libmv_image);

  if (mask) {
    libmv_floatBufferToFloatImage(mask, width, height, 1, &libmv_mask);
    libmv_mask_for_sample = &libmv_mask;
  }

  SamplePlanarPatch(libmv_image,
                    xs, ys,
                    num_samples_x, num_samples_y,
                    libmv_mask_for_sample,
                    &libmv_patch,
                    warped_position_x,
                    warped_position_y);

  libmv_floatImageToFloatBuffer(libmv_patch, patch);
}

void libmv_samplePlanarPatchByte(const unsigned char* image,
                                  int width,
                                  int height,
                                  int channels,
                                  const double* xs,
                                  const double* ys,
                                  int num_samples_x,
                                  int num_samples_y,
                                  const float* mask,
                                  unsigned char* patch,
                                  double* warped_position_x,
                                  double* warped_position_y) {
  libmv::FloatImage libmv_image, libmv_patch, libmv_mask;
  libmv::FloatImage *libmv_mask_for_sample = NULL;

  libmv_byteBufferToFloatImage(image, width, height, channels, &libmv_image);

  if (mask) {
    libmv_floatBufferToFloatImage(mask, width, height, 1, &libmv_mask);
    libmv_mask_for_sample = &libmv_mask;
  }

  libmv::SamplePlanarPatch(libmv_image,
                           xs, ys,
                           num_samples_x, num_samples_y,
                           libmv_mask_for_sample,
                           &libmv_patch,
                           warped_position_x,
                           warped_position_y);

  libmv_floatImageToByteBuffer(libmv_patch, patch);
}
